### PR TITLE
Remove duplicates in the user_ips table and add an index

### PR DESCRIPTION
**Pull request by hawkowl**

### CODE REVIEW

**Comment by erikjohnston** [comment]
A couple of notes:

1. Given how large this table is I'd probably want to do this using the batching mechanism (which correctly balances the batches so that it doesn't overload the DB)
2. We probably want to have the index be `(user_id, access_token, ip)`, i.e. remove `user_agent` and add `user_id`. We don't really care about keeping unique UA's, we just use them to have an idea about what type the device is, and it really shouldn't change much in practice anyway. We add `user_id` mainly so that we can reuse the existing index during the de-duplication process (and then we can drop the existing index after adding the new one)
3. We need to be careful that new inserts don't add duplicates while we're de-duplicating.

The way I'd probably do this is by iterating over the table using `last_seen`, and then deleting rows that have duplicate entries. Something like:

1. Find a span of last_seen times that have roughly the right number of rows:
```sql
SELECT last_seen FROM user_ips
WHERE last_seen > ?
ORDER BY last_seen 
LIMIT 1
OFFSET ?
```
2. Fetch all rows in that span which have duplicate entries (I've checked this query is fast enough):
```sql
SELECT user_id, access_token, ip, MAX(last_seen)
FROM (
    SELECT user_id, access_token, ip
    FROM user_ips
    WHERE ? <= last_seen AND last_seen < ?
    ORDER BY last_seen
) c
INNER JOIN user_ips USING (user_id, access_token, ip)
GROUP BY user_id, access_token, ip;
HAVING count(*) > 1
```

3. Delete all rows matching the returned `(user_id, access_token, ip)` and reinserting one row with the returned last_seen.

4. Repeat.

After we've successfully walked the entire table we probably need to:

1. Stop inserting into user_ips
2. Rerun the deduplicate for the time span since we started deduplication, i.e. for the period we might have inserted into
3. Create the unique index.
4. Allow insertions again.

Otherwise we risk inserting duplicates while trying to build the unique index :(

**Comment by hawkowl** [comment]
The conversation that led to the indexing changes: https://matrix.to/#/!yZHTGeDKZUeKaqeTeU:matrix.org/$1546981086143427NlIBG:matrix.org?via=matrix.org&via=sw1v.org&via=meshspace.de&via=half-shot.uk

File: synapse/storage/client_ips.py [hunk 1]
```diff
@@ -75,6 +89,86 @@ def __init__(self, db_conn, hs):
...
+
+                # Add in one to be the last_seen
+                txn.execute(
+                    """
+                    INSERT INTO user_ips
+                    (user_id, access_token, ip, device_id, user_agent, last_seen)
+                    VALUES (?, ?, ?, ?, ?, ?)
+                    """,
+                    (user_id, access_token, ip, device_id, user_agent, last_seen)
+                )
```
Comment (erikjohnston) [inline synapse/storage/client_ips.py]: May as well use the `_simple_{insert,delete}_txn` wrappers here?

File: synapse/storage/client_ips.py [hunk 2]
```diff
@@ -75,6 +89,86 @@ def __init__(self, db_conn, hs):
...
+                    FROM user_ips
+                    WHERE ? <= last_seen AND last_seen < ?
+                    ORDER BY last_seen
+                ) c
+                INNER JOIN user_ips USING (user_id, access_token, ip)
+                GROUP BY user_id, access_token, ip;
+                HAVING count(*) > 1""",
+                (last_seen_progress, last_seen)
+            )
+            res = txn.fetchall()
```
Comment (erikjohnston) [inline synapse/storage/client_ips.py]: Some quick words about what the fuck this query is doing might be something that other people might appreciate

File: synapse/storage/client_ips.py (hunk 1 above)
Comment (hawkowl) [reply synapse/storage/client_ips.py]: yeahhhh but everything else in this function is sql :P

File: synapse/storage/client_ips.py (hunk 2 above)
Comment (erikjohnston) [reply synapse/storage/client_ips.py]: Something like:

> This works by looking at all entries in the given time span, and then for each `(user_id, access_token, ip)` tuple in that range checking for any duplicates in the rest of the table (via a join). It then only returns entries which have duplicates, and the max `last_seen` across all duplicates, which can the be used to delete all other duplicates.  
> It is efficient due to the existence of  `(user_id, access_token, ip)` and `(last_seen)` indices.

File: synapse/storage/client_ips.py [hunk 3]
```diff
@@ -169,8 +289,8 @@ def get_last_client_ip_by_device(self, user_id, device_id):
 
         ret = {(d["user_id"], d["device_id"]): d for d in res}
         for key in self._batch_row_update:
-            uid, access_token, ip = key
-            if uid == user_id:
+            user_id, access_token, ip = key
+            if user_id == user_id:
```
Comment (erikjohnston) [inline synapse/storage/client_ips.py]: *cough*

Comment (hawkowl) [reply synapse/storage/client_ips.py]: HMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMM 

yeah need some coffee

File: synapse/storage/client_ips.py [hunk 4]
```diff
@@ -75,6 +95,106 @@ def __init__(self, db_conn, hs):
...
+                SELECT user_id, access_token, ip,
+                       MAX(device_id), MAX(user_agent), MAX(last_seen)
+                FROM (
+                    SELECT user_id, access_token, ip
+                    FROM user_ips
+                    WHERE ? <= last_seen AND last_seen < ?
+                    ORDER BY last_seen
+                ) c
+                INNER JOIN user_ips USING (user_id, access_token, ip)
+                GROUP BY user_id, access_token, ip;
```
Comment (erikjohnston) [inline synapse/storage/client_ips.py]: Spurious railing `;`

File: synapse/storage/client_ips.py [hunk 5]
```diff
@@ -75,6 +95,106 @@ def __init__(self, db_conn, hs):
...
+                txn.execute(
+                    """
+                    INSERT INTO user_ips
+                    (user_id, access_token, ip, device_id, user_agent, last_seen)
+                    VALUES (?, ?, ?, ?, ?, ?)
+                    """,
+                    (user_id, access_token, ip, device_id, user_agent, last_seen)
+                )
+
+            self._background_update_progress_txn(txn, {"last_seen": last_seen})
```
Comment (erikjohnston) [inline synapse/storage/client_ips.py]: Missing update name param

**Pull request merged**